ql/src/java/org/apache/hadoop/hive/ql/optimizer/UnionDistinctMerger.java

```diff
@@ -0,0 +1,184 @@
+      GroupByOperator lowerFinalGroupByOperator = (GroupByOperator) stack.get(stack.size() - 1);
+
+      if (upperFinalGroupByOperator.getConf().getAggregators().isEmpty() &&
+          lowerFinalGroupByOperator.getConf().getAggregators().isEmpty()) {
+        LOG.info("Detect duplicate UNION-DISTINCT GBY patterns. Remove the latter one.");
+
+        // Step 0. UNION1->GBY1->RS1->GBY2->UNION2->GBY3->RS2->GBY4
+
+        // Step 1. Cut GBY2->UNINO1
+        lowerUnionOperator.removeParent(upperFinalGroupByOperator);
+
+        // Step 2. Redirect the output of the parents of lowerUnionOperator to upperUnionOperator.
+        // Before step 2:
+        //    OP1-UNION1->GBY1->RS1->GBY2
+        //    OP2-
+        //    OP3-UNION2->GBY3->RS2->GBY4
+        //    OP4-
+        // After step 2:
+        //    OP1-UNION1->GBY1->RS1->GBY2
+        //    OP2-
+        //    OP3-
+        //    OP4-
+        for (Operator<?> lowerUnionParent: lowerUnionOperator.getParentOperators()) {
+          lowerUnionParent.replaceChild(lowerUnionOperator, upperUnionOperator);
+          upperUnionOperator.getParentOperators().add(lowerUnionParent);
+        }
+        lowerUnionOperator.setParentOperators(new ArrayList<>());
+
+        // Step 3. Redirect the output of lowerFinalGroupByOperator to children of
+        // Before step 3:
+        //    OP1-UNION1->GBY1->RS1->GBY2
+        //    OP2-
+        //    OP3-UNION2->GBY3->RS2->GBY4
+        //    OP4-
+        // After step 3:
+        //    OP1-UNION1->GBY1->RS1->GBY2
+        //    OP2-
+        //    OP3-
+        //    OP4-
+        for (Operator<?> lowerFinalGroupByChild: lowerFinalGroupByOperator.getChildOperators()) {
+          lowerFinalGroupByChild.replaceParent(lowerFinalGroupByOperator, upperFinalGroupByOperator);
+          upperFinalGroupByOperator.getChildOperators().add(lowerFinalGroupByChild);
+        }
+        upperUnionOperator.getConf().setNumInputs(upperUnionOperator.getNumParent());
+      }
+
+      return null;
+    }
+  }
+
+  private static class NoSkipGraphWalker extends DefaultGraphWalker {
+    public NoSkipGraphWalker(SemanticDispatcher disp) {
+      super(disp);
+    }
+
+    public void startWalking(Collection<Node> startNodes,
+        HashMap<Node, Object> nodeOutput) throws SemanticException {
+      toWalk.addAll(startNodes);
+      while (toWalk.size() > 0) {
+        Node nd = toWalk.remove(0);
+        walk(nd);
+        // We need to revisit GroupBy operator for every distinct operator path.
+        // GraphWalker uses retMap to determine if an operator has been visited.
+        // Clearing it after each walk() ensures that we visit GroupBy operator in every possible path.
+        retMap.clear();
+      }
+    }
+  }
+
+  public ParseContext transform(ParseContext pCtx) throws SemanticException {
+    Map<SemanticRule, SemanticNodeProcessor> testRules = new LinkedHashMap<>();
+    StringBuilder pattern = new StringBuilder();
```

Comment:
do we need to construct this pattern every time? can we extract it to constants?

Reply:
Moved the pattern to outside of constructor.